### PR TITLE
Metrics v2 first steel thread with a LongCounter

### PR DESCRIPTION
**Pull request by jkwatson**
If there is appetite for #2033, here is the very first "steel thread" that implements a LongCounter from end-to-end.

There are obviously tons of todos here, and lots of work to be done before this could be considered ready to use.

### CODE REVIEW

**Comment by carlosalberto** [comment]
Nice!

**Review by thisthat — COMMENTED** [review]
Thank you for taking care of this! I also think the metric SDK needs to be rewritten to be better aligned with the specification. I wonder if we should have this as an extension rather than an SDK package at the moment. Or maybe adding "beta" or other labels to the module name could also discourage users to think this is fully working and a replacement of the current metric SDK.

File: sdk/metrics-v2/build.gradle [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    testAnnotationProcessor libraries.auto_value
+    testCompileOnly libraries.auto_value_annotation
+
+    testCompile project(path: ':opentelemetry-sdk-common', configuration: 'testClasses')
+
+//    testImplementation project(':opentelemetry-sdk-testing')
+    testImplementation libraries.junit_pioneer
+
+    signature "net.sf.androidscents.signature:android-api-level-24:7.0_r2@signature"
```
Comment (anuraaga) [inline sdk/metrics-v2/build.gradle:27]: gummy bears plz

**Review by anuraaga — COMMENTED** [review]
@jkwatson Thanks for working on this - any recommendation on an ordering to review this PR in for better general context? Or just a line-by-line java review is enough?

**Comment by jkwatson** [comment]
> @jkwatson Thanks for working on this - any recommendation on an ordering to review this PR in for better general context? Or just a line-by-line java review is enough?

I think I'd start by looking at the in-process SDK spec, and looking at the general structure. The line-by-line at this point is less important than the general architecture of the solution matching what is being specified. I would guess that most of the lines will end up being replaced at some point, as we evolve the solution towards being production ready.

**Comment by jkwatson** [comment]
> Thank you for taking care of this! I also think the metric SDK needs to be rewritten to be better aligned with the specification. I wonder if we should have this as an extension rather than an SDK package at the moment. Or maybe adding "beta" or other labels to the module name could also discourage users to think this is fully working and a replacement of the current metric SDK.

Any and all suggestions welcome. At the moment, this isn't wired into the default SDK, so it would only end up being used if someone specifically added this module as a dependency to their application setup. Maybe we could not publish it anywhere until its ready, so no one could depend on it?

File: sdk/metrics-v2/build.gradle (hunk 1 above)
Comment (jkwatson) [reply sdk/metrics-v2/build.gradle]: done

**Comment by jkwatson** [comment]
> > @jkwatson Thanks for working on this - any recommendation on an ordering to review this PR in for better general context? Or just a line-by-line java review is enough?
> 
> I think I'd start by looking at the work-in-progress SDK spec, and looking at the general structure. The line-by-line at this point is less important than the general architecture of the solution matching what is being specified. I would guess that most of the lines will end up being replaced at some point, as we evolve the solution towards being production ready.

**Comment by jkwatson** [comment]
I'm going to close this for now, pending work on morphing our existing SDK to look more like the spec.

**Pull request closed without merging**